### PR TITLE
PoC: server handling multiple clients with custom attention mask api

### PR DESCRIPTION
**Pull request by FSSRepo**
From #3462, I wanted to update my fork to the latest changes from the master branch, but it went wrong :(.

Hello, I know it's something no one asked for, but some of us need it. Here's a proof of concept of a server that handles multiple clients simultaneously, thanks to the new way of working with the KV cache.

Some may wonder why reimplementing this proposal in a separate example. The current implementation of the server is quite complex, and many things could break.

Tested on:
```bash
Windows 11 x64
Intel Core i5 11400H 6 C / 12 T
RTX 3050 laptop 4 GB VRAM
16 GB of RAM DDR4 3200MHz
````

https://github.com/ggerganov/llama.cpp/assets/57494570/cb867d2d-f4b1-4b53-85c4-94ce47532fbc

This is a proof of concept for now, with some feedback and assistance, we could make it more usable.

Here is the command to start the server:

```bash
./server-parallel -m models/7B/ggml-model.gguf --ctx_size 2048 -t 4 -ngl 33 --batch-size 512 --parallel 3 -n 512 --cont-batching
```

Modify `--parallel` to the number of slots to process clients requests.

Edit:
- New video showing 4 clients at same time, my laptop almost exploded 😂.
- Improve the PR note.

Lastest changes:
- Improved README
- Use a custom system prompt and change it on runtime

Note:

Many people are going to want to kill me when they see how I handle multithreading without using mutex; I never knew what that was for :(.

### CODE REVIEW

**Comment by ggerganov** [comment]
Works great! Here is another demo on M1 Pro with F16 7B:

https://github.com/ggerganov/llama.cpp/assets/1991296/0ad5f642-93e4-4899-93cd-4a60d6b36dd8

@FSSRepo Can you allow pushes to your branch so I can push some fixes:

```
$ ▶ git push FSSRepo HEAD:fixes 
error: Authentication error: Authentication required: You must have push access to verify locks
error: failed to push some refs to 'https://github.com/FSSRepo/llama.cpp'
```

File: examples/server-parallel/server.cpp [hunk 1]
```diff
@@ -0,0 +1,884 @@
+#include "frontend.h"
+#include "common.h"
+#include "llama.h"
+
+#include "../server/httplib.h"
+#include "../server/json.hpp"
```
Comment (ggerganov) [inline examples/server-parallel/server.cpp:6]: Should move these two headers from `/examples/server` into `/common`

Comment (FSSRepo) [reply examples/server-parallel/server.cpp]: Done!

**Comment by FSSRepo** [comment]
> @FSSRepo Can you allow pushes to your branch so I can push some fixes:

Screenshot 2023-10-06 094305

I don't have some restriction in that branch

**Comment by kiratp** [comment]
I am going to take the shameless opportunity here to request that adding support for speculative execution be considered here - that would make the first and only OSS LLM server I have come across that supports it out of the box!

**Comment by Seikho** [comment]
Can this support the same generation parameters that are used in the existing example server?

**Comment by ggerganov** [comment]
Here is a 30B LLaMA Q4_0 serving 4 clients on M2 Ultra:

https://github.com/ggerganov/llama.cpp/assets/1991296/c1d0f138-2f48-4b26-adef-471e2544b7f9

*(I tried 7B and 13B, but the generation is so fast that I am not able to start the requests in parallel)*

**Comment by FSSRepo** [comment]
> *(I tried 7B and 13B, but the generation is so fast that I am not able to start the requests in parallel)*

I want to add the option to cancel the streaming but when I use AbortControl in the frontend it causes error, I'm thinking add a get endpoint `stop?slot=id` to request release the sequence.

**Comment by ggerganov** [comment]
And one more example using the original `server` UI with 70B Q8_0 model:

https://github.com/ggerganov/llama.cpp/assets/1991296/4205bc28-c769-4694-8b29-7a39ddb19340

**Comment by FSSRepo** [comment]
> And one more example using the original `server` UI with 70B Q8_0

Are you working in implement that ui or just reuse the endpoints?

**Comment by ggerganov** [comment]
> Are you working in implement that ui or just reuse the endpoints?

No, I just noticed it works and have it a try. I thought you implemented it.

> I want to add the option to cancel the streaming but when I use AbortControl in the frontend it causes error, I'm thinking add a get endpoint stop?slot=id to request release the sequence.

What kind of error?

**Comment by AdityaSher** [comment]
How is this working? Are the instances sharing the same weights or the model needs to be loaded N times?

**Comment by FSSRepo** [comment]
> How is this working? Are the instances sharing the same weights or the model needs to be loaded N times?

Just divides the kv cache (context size) to a number of sequences, the limit is the context size because it is shared between clients.

This not reload the model, just once.

**Comment by FSSRepo** [comment]
> No, I just noticed it works and gave it a try. I thought you implemented it.

Only had reused the completion.js function. 😂

**Comment by FSSRepo** [comment]
> What kind of error?
Screenshot 2023-10-08 173203

```javascript
 controller = new AbortController();
    const response = await fetch("http://localhost:8080/completion", {
      method: "POST",
      body: JSON.stringify(options),
      headers: {
        Connection: "keep-alive",
        "Content-Type": "application/json",
        Accept: "text/event-stream",
      },
      signal: controller.signal,
    });
```

function cancel() {
  if(controller) {
    controller.abort(); // when i call this function i get DOMException
  }
}

**Comment by KerfuffleV2** [comment]
> when I use AbortControl in the frontend it causes error,

"*Note: When `abort()` is called, the `fetch()` promise rejects with a `DOMException` named `AbortError`.*" — https://developer.mozilla.org/en-US/docs/Web/API/AbortController

So that's probably normal. You can possibly try to catch the exception, like in the example there. I wouldn't think that would cause the connection not to abort properly though, so it not stopping generation on the server side is probably a different problem.

**Comment by FSSRepo** [comment]
> so it not stopping generation on the server side is probably a different problem.

I will fix it adding a GET endpoint `stop_completion` to notify the server release the slot.

**Comment by FSSRepo** [comment]
@ggerganov can you merge the lastest changes of master in this branch, please. I'm afraid to make a mistake again.

File: examples/server-parallel/server.cpp [hunk 2]
```diff
@@ -0,0 +1,939 @@
...
+                        slot->sampled_tokens.pop_back();
+                        ss << "data: " << res_d.dump() << "\n\n";
+                        string result = ss.str();
+                        if(!sink.write(result.c_str(), result.size())) {
+                            slot->release();
+                            return false;
+                        }
+                    }
+                    return true;
+                });
```
Comment (jhen0409) [inline examples/server-parallel/server.cpp:905]:
```suggestion
                },
                [slot](bool) {
                    slot->release();
                });
```

For abort issue, I think it missing a complete handler.

Comment (FSSRepo) [reply examples/server-parallel/server.cpp]: This works!! thank you for this idea 😊

**Comment by FSSRepo** [comment]
> This example is very nice to have, but I'm thinking if we should try to directly implement the functionality in the existing `server` example. I'm not sure if I'll have time to do it myself though and I don't want to delay this super long. Also, the current implementation keeps one CPU core at 100% all the time which is not desirable.
> 
> I'll give this PR some more time to see if people would be interested in improving this. If not, we will probably merge this and add an item on the roadmap for the future.
> 

I will do it

**Comment by FSSRepo** [comment]
@cebtenzzre Thank you so much!

**Comment by FSSRepo** [comment]
@ggerganov "Do you want me to close this PR? I initially proposed this as a simple example to avoid the complexity of the server example.

**Comment by ggerganov** [comment]
@FSSRepo Let's see if we can make the other PR work and if so, we will close this one

**Comment by ggerganov** [comment]
Superseded #3677

**Pull request closed without merging**